### PR TITLE
Intermediate.ATX.2017Q1.Ava.Gore

### PR DESCRIPTION
**Pull request by avagore**
# Checkpoint Rubric

This is the rubric that your instructor will use to grade your checkpoints. Please do not edit. 

## Checkpoint 1: Towers of Hanoi Terminal App
- [x] All tests passed: 55 points
- [x] Properly commented code: 15 points
- [x] Properly indented code: 15 points
- [x] Demonstrated effective use of JavaScript: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

## Checkpoint 2: Towers of Hanoi GUI App
- [x] All tests passed: 55 points
- [x] Properly commented code: 15 points
- [x] Properly indented code: 15 points
- [x] Demonstrated effective use of jQuery: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

## Checkpoint 3: Final Project Gist Blog
- [x] Effective use of AJAX: 20 points
- [x] Effective use of loops: 20 points
- [x] Effective use of github API: 20 points
- [x] Style and design: 10 points
- [x] Properly commented code: 15 points
- [x] Properly indented code: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

### CODE REVIEW

File: apps/02RockPaperScissors.js [hunk 1]
```diff
@@ -9,13 +9,6 @@ var rl = readline.createInterface({
 
 // hand1 and hand2 are variables that will take the values: 'rock', 'paper', or 'scissors'.
 function rockPaperScissors(hand1, hand2) {
-  if(hand1.toLowerCase() !== 'rock' || 'paper' || 'scissors') {
-    return "Could not recognize input. Please enter either 'rock', 'paper', or 'scissors'.";
-  }
-    else if(hand2.toLowerCase() !== 'rock' || 'paper' || 'scissors') {
```
Comment (athensbird) [inline apps/02RockPaperScissors.js]: On line 15, indent it with the if the statement

File: apps/02RockPaperScissors.js [hunk 2]
```diff
@@ -43,7 +36,7 @@ function rockPaperScissors(hand1, hand2) {
     }
   };
 
-  
+
```
Comment (athensbird) [inline apps/02RockPaperScissors.js]: Get rid of the extra spaces.

File: apps/02RockPaperScissors.js [hunk 3]
```diff
@@ -9,13 +9,6 @@ var rl = readline.createInterface({
```
Comment (athensbird) [inline apps/02RockPaperScissors.js:9]: Good job overall. Neat indentation and alignment.

File: apps/03TicTacToe.js [hunk 4]
```diff
@@ -22,25 +22,70 @@ function printBoard() {
   console.log('  ---------');
   console.log('2 ' + board[2].join(' | '));
 }
+//starting a new game after a win
+function newGame() {
+  board = [
+    [' ', ' ', ' '],
+    [' ', ' ', ' '],
+    [' ', ' ', ' ']
+  ];
```
Comment (pieroapretto) [inline apps/03TicTacToe.js:31]: It's good practice to return a value in every function.

In this instance, you can `return board;`

File: apps/03TicTacToe.js [hunk 5]
```diff
@@ -22,25 +22,70 @@ function printBoard() {
...
 
 function ticTacToe(row, column) {
-  // Your code here
+  // checks for invalid input
+  if(row > 2 || isNaN(row)) {
+    console.log("Invalid input.");
+    return false;
+  } else if(column > 2 || isNaN(column)) {
+    console.log("Invalid input.");
+    return false;
```
Comment (pieroapretto) [inline apps/03TicTacToe.js:75]: Great data scrubbing!

File: apps/03TicTacToe.js [hunk 6]
```diff
@@ -22,25 +22,70 @@ function printBoard() {
...
+  if((board[0][0] === playerTurn && board[1][1] === playerTurn && board[2][2] === playerTurn) ||
+  (board[0][2] === playerTurn && board[1][1] === playerTurn && board[2][0] === playerTurn)) {
+    return true;
+  }
 }
-
+//checking if a player has won
 function checkForWin() {
-  // Your code here
+  if(horizontalWin() === true || verticalWin() === true || diagonalWin() === true) {
```
Comment (pieroapretto) [inline apps/03TicTacToe.js]: Declaring a function in an `if` statement will automatically evaluate the 'truthiness' of that function.

You can shorthand your code doing the following:

`if(horizontalWin() || verticalWin() || diagonalWin()) { ...`

File: homework/03LessonThree.js [hunk 7]
```diff
@@ -110,10 +112,13 @@ var threeByThreeArray = [
 
 // Problem 11:
 // Replace the center element of threeByThreeArray, which is 5, with something else.
-
+threeByThreeArray[1][1] = 8;
```
Comment (pieroapretto) [inline homework/03LessonThree.js:115]: NICE!

File: homework/04LessonFour.js [hunk 8]
```diff
@@ -77,6 +87,7 @@ var nestedObjects = {
   }
 };
 // update the starring property to have the value 'Tim Robbins'
+nestedObjects['users']['bob roberts']['starring'] = 'Tim Robbins';
```
Comment (pieroapretto) [inline homework/04LessonFour.js:90]: nicelydone

File: apps/03TicTacToe.js [hunk 9]
```diff
@@ -22,25 +22,70 @@ function printBoard() {
...
 function ticTacToe(row, column) {
-  // Your code here
+  // checks for invalid input
+  if(row > 2 || isNaN(row)) {
+    console.log("Invalid input.");
+    return false;
+  } else if(column > 2 || isNaN(column)) {
+    console.log("Invalid input.");
+    return false;
+  }
```
Comment (ASTPMeetup) [inline apps/03TicTacToe.js]: Nice data scrubbing.

File: apps/04TowersOfHanoi.js [hunk 10]
```diff
@@ -13,30 +13,92 @@ var stacks = {
...
+  var endLength = stacks[endStack].length;
+  //if the end stack is empty, any piece can go there.
+  if(stacks[endStack].length < 1) {
+    return true;
+  //if the piece being moved is smaller than the piece on the end stack, it is a legal move.
+  } else if((stacks[startStack][startLength - 1]) < (stacks[endStack][endLength - 1])) {
+      if(startStack !== 'a' || startStack !== 'b' || startStack !== 'c') {
+        console.log("Invalid input. Try again.");
+        return false;
+      } else if(startStack !== isNaN) {
```
Comment (pieroapretto) [inline apps/04TowersOfHanoi.js]: Consolidate your `if` and `else if` statements with another logical operator.

File: apps/04TowersOfHanoi.js [hunk 11]
```diff
@@ -13,30 +13,92 @@ var stacks = {
...
 
 }
 
 function towersOfHanoi(startStack, endStack) {
-  // Your code here
 
+  startStack = startStack.toLowerCase();
+  endStack = endStack.toLowerCase();
+//if the move is legal, move the piece
+  if(isLegal(startStack, endStack) === true) {
```
Comment (pieroapretto) [inline apps/04TowersOfHanoi.js]: You can pass in the function directly and the `if` statement will only run if the function returns a 'truthy' value.

example: `if(returnTrueOrFalse()) { return "I am true!" }`

File: apps/04TowersOfHanoi.js [hunk 12]
```diff
@@ -13,30 +13,92 @@ var stacks = {
...
+    movePiece(startStack, endStack);
+    // Prints out how many moves user makes.
+    console.log("Number of moves: " + counter);
+    checkForWin();
```
Comment (pieroapretto) [inline apps/04TowersOfHanoi.js:82]: All tests passed!

dance

File: apps/06.1Checkers.js [hunk 13]
```diff
@@ -8,12 +8,53 @@ var rl = readline.createInterface({
...
+      //spot is an array
+      //row, column [0, 1]
+      var row = spot[0];
+      var column = spot[1];
+      var whiteChecker = new Checker('white');
+      this.grid[row][column] = whiteChecker;
+      this.checkers.push(whiteChecker);
+    }
+
+    for(var spot of blackPositions) {
```
Comment (pieroapretto) [inline apps/06.1Checkers.js:50]: Neat for loop.

File: apps/06.1Checkers.js [hunk 14]
```diff
@@ -53,15 +94,46 @@ function Board() {
...
+    this.board.grid[startRow][startColumn] = null;
+    this.board.grid[endRow][endColumn] = checker;
+    
+    var distance = Math.abs(startRow - endRow);
+    var midRow = (startRow + endRow) / 2;
+    var midCol = (startColumn + endColumn) / 2;
+    var killPosition = [midRow, midCol];
+
+    if(distance === 2) {
+      this.board.killChecker(killPosition);
```
Comment (pieroapretto) [inline apps/06.1Checkers.js:133]: All tests passed!

wizard

File: apps/06SpaceTravelToMars.js [hunk 15]
```diff
@@ -9,7 +9,51 @@ var jobTypes = {
   programmer: 'Any Ship!'
 };
 
+
 // Your code here
+class CrewMember {
+  constructor(name, job, specialSkill) {
+    if(name === '') {
```
Comment (pieroapretto) [inline apps/06SpaceTravelToMars.js:16]: Good data scrubbing.

File: apps/06SpaceTravelToMars.js [hunk 16]
```diff
@@ -9,7 +9,51 @@ var jobTypes = {
...
+  constructor(name, type, ability) {
+    this.name = name;
+    this.type = type;
+    this.ability = ability;
+    this.crew = [];
+
+    this.missionStatement = function() {
+      var missionStatus = "Can't perform a mission yet.";
+      for (var i = 0; i < this.crew.length; i++) {
+        if (jobTypes[this.crew[i].job] === this.type || this.crew[i].job === 'programmer') {
```
Comment (pieroapretto) [inline apps/06SpaceTravelToMars.js:42]: ` this.crew[i].job === 'programmer`??

File: homework/06LessonSix.js [hunk 17]
```diff
@@ -76,28 +81,32 @@ function ConferenceRoom() {
...
 
 // ****
 // Concept Checkpoint
 //
 // Write your answer in comments
 //
 // What is a method?
-//
-//
+// A method is a function that contains a function definition. It is used to add definitions
```
Comment (pieroapretto) [inline homework/06LessonSix.js:108]: A method is a _property_ that contains a function definition.

It's a function that is assigned to a object as a object property.

File: apps/07jquery-tic-tac-toe/script.js [hunk 18]
```diff
@@ -1,5 +1,72 @@
...
+var boxSeven = $("[data-cell='7']");
+var boxEight = $("[data-cell='8']");
+
 $(document).on('ready', function() {
-  // Put app logic in here
+  $('[data-cell]').click(function() {
+    $(this).text(playerTurn);
+    checkForWin();
+    playerTurn = (playerTurn === 'X') ? 'O' : 'X';
+  })
```
Comment (pieroapretto) [inline apps/07jquery-tic-tac-toe/script.js:19]: Missing semicolon on line 19.

File: apps/07jquery-tic-tac-toe/script.js [hunk 19]
```diff
@@ -1,5 +1,72 @@
...
+
+  $('#clear').click(function() {
+    $('[data-cell]').text("");
+    $('#announce-winner').text("");
+    playerTurn = 'X';
+  })
+
+
+  function horizontalWin() {
+    if((boxZero.text() === playerTurn) && (boxOne.text() === playerTurn) && (boxTwo.text() === playerTurn)) {
```
Comment (pieroapretto) [inline apps/07jquery-tic-tac-toe/script.js:29]: Much easier to read and manage with the jQuery objects stored in variables.

Nice work.

File: homework/07LessonSeven/script.js [hunk 20]
```diff
@@ -69,20 +69,23 @@ $(document).on('ready', function() {
...
   // and set window.dataIdValue to the value of the data-id attribute of the element
   // you can do this by calling .data() on the selected element
-  window.dataIdValue;
+  window.dataIdValue = $('#data-element').data('id');
 
   // Problem 12:
   // select the element with id="data-element"
   // and set its text (.text()) to the value of the data-content attribute
+  var content = $('#data-element').data('content');
+  $('#data-element').text(content);
```
Comment (pieroapretto) [inline homework/07LessonSeven/script.js:90]: All tests passed!

nailedit

**Pull request closed without merging**